cmake:bugfix generate config.h contain ; characters will be handled incorrectly

xuxin930 opened the pull request:
## Summary
### Issue
`;` is treated as a list separator in CMake.
when setting a configuration similar to the following,
an error will occur when generating config.h
```
CONFIG_LV_TXT_BREAK_CHARS " ,.;:-_)]}"
```
> [!CAUTION]
> This header file is completely wrong from line 237 onwards
![img_v2_cfa97d61-86f3-43ff-b771-3be50c2602dl](https://github.com/apache/nuttx/assets/31881278/38687fbe-f00c-4b52-bd47-a0ffdd3bf2a7)

### Reason

This is because the file(STRING) function will read the wrong .config value line.
and  string(REGEX REPLACE) will also incorrectly handle lines containing `;` .

### Solution

I changed the reading process to parse the entire .config as a big string.
parse each line character by character.
when there is a special character `;` , it is treated as a list and finally spliced.

## Impact

## Testing
CI build

xuxin930 commented:
Hi @anchao @XuNeo 
please help review whether this is correct 🙏

XuNeo commented:
I have tested with lvgl v9 and it works!

acassis commented:
@xuxin930 nice description of a problem and its solution! Kudos!!!

anchao commented:
Is there any other way? This implementation seems complex and ugly，have you tried this config in zephyr build system？

xiaoxiang781216 commented on cmake/nuttx_mkconfig.cmake:
```diff
@@ -91,13 +91,21 @@ file(APPEND ${CONFIG_H}
-foreach(NameAndValue ${ConfigContents})
+function(process_individual_rows)
+  set(NameAndValue ${ARGN})
+  list(LENGTH NameAndValue LENGTH_RESULT)
```
let's keep the naming style consistent (NameAndValue, LENGTH_RESULT, configFileContent).

anchao commented:
This issue not caused by `';'` character, which should be `foreach()` treated close bracket ` ']' ` as a regular expression:
https://stackoverflow.com/questions/26057678/escape-characters-in-a-semicolon-separated-list-in-cmake
https://cmake.org/pipermail/cmake/2016-March/063030.html

@xuxin930 try this patch to check if there is any improvement.
```
diff --git a/cmake/nuttx_mkconfig.cmake b/cmake/nuttx_mkconfig.cmake
index fc040eb840..88353e2042 100644
--- a/cmake/nuttx_mkconfig.cmake
+++ b/cmake/nuttx_mkconfig.cmake
@@ -92,7 +92,11 @@ file(APPEND ${CONFIG_H}
 file(APPEND ${CONFIG_H} "#define CONFIG_BASE_DEFCONFIG \"${BASE_DEFCONFIG}\"\n")
 
 file(STRINGS ${CMAKE_BINARY_DIR}/.config ConfigContents)
+STRING(REGEX REPLACE "\\[" "__OPEN_BRACKET__" ConfigContents "${ConfigContents}")
+STRING(REGEX REPLACE "\\]" "__CLOSE_BRACKET__" ConfigContents "${ConfigContents}")
 foreach(NameAndValue ${ConfigContents})
+  STRING(REGEX REPLACE "__OPEN_BRACKET__" "[" NameAndValue "${NameAndValue}")
+  STRING(REGEX REPLACE "__CLOSE_BRACKET__" "]" NameAndValue "${NameAndValue}")
   string(REGEX REPLACE "^[ ]+" "" NameAndValue ${NameAndValue})
   string(REGEX MATCH "^CONFIG[^=]+" NAME ${NameAndValue})
   # skip BASE_DEFCONFIG here as it is handled above

```

xuxin930 commented:
> This issue not caused by `';'` character, which should be `foreach()` treated close bracket `']'` as a regular expression: https://stackoverflow.com/questions/26057678/escape-characters-in-a-semicolon-separated-list-in-cmake https://cmake.org/pipermail/cmake/2016-March/063030.html
> 
> @xuxin930 try this patch to check if there is any improvement.
> 

Yes! 👍
I changed the solution and modified other methods of reading config to ensure consistency.
Unfortunately🙁, CMake does NOT provide calls like **proxy functions** or **wrapper function**s. 
I can only use marco to replace them.

xuxin930 commented:
> @xuxin930 nice description of a problem and its solution! Kudos!!!

@acassis thank you very much for your encouragement, I will continue to be as clear as possible when dealing with issues
